middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+        deps: { keyStore },
+        masterAccount: creatorKeyJson && creatorKeyJson.account_id,
+        nodeUrl: process.env.NODE_URL
+    });
+    const contractAccount = new nearAPI.Account(near.connection, contractName);
+    const contract = new nearAPI.Contract(contractAccount, contractName, {
+        viewMethods,
+        changeMethods,
+    });
+    return contract;
+};
+const confirmRequest = async (accountId, request_id) => {
+    const key = await getDetermKey(accountId);
+    const contract = await getContract(accountId, key.secretKey);
+    // always parseInt on requestId. requestId is string in db because it could come from url params etc...
+    const res = await contract.confirm({ request_id: parseInt(request_id) }).catch((e) => {
+        return { success: false, error: e };
+    });
+    return { success: !!res, res };
+};
+/********************************
+2FA Key
+********************************/
+// helper: generates a deterministic key based on the accountId
+const getDetermKey = async (accountId) => {
+    const hash = crypto.createHash('sha256').update(accountId + creatorKeyJson.private_key).digest();
+    const keyPair = nacl.sign.keyPair.fromSeed(hash);//nacl.sign.keyPair.fromSecretKey(hash)
+    return {
+        publicKey: `ed25519:${nearAPI.utils.serialize.base_encode(keyPair.publicKey)}`,
+        secretKey: nearAPI.utils.serialize.base_encode(keyPair.secretKey)
+    };
+};
+/********************************
+Sending Codes
+method.kind = ['2fa-email', '2fa-phone']
+********************************/
+const prettyRequestInfo = ({ request }) => `
+    Transaction Recipient: ${ request.receiver_id }
+    Actions:\n\t${ request.actions.map((r) => r.type + (r.amount ? ': ' + nearAPI.utils.format.formatNearAmount(r.amount, 4) : '')).join('\n\t') }
+`;
+
+const sendCode = async (method, twoFactorMethod, requestId = -1, data = {}) => {
+    const securityCode = password.randomPassword({ length: SECURITY_CODE_DIGITS, characters: password.digits });
+    await twoFactorMethod.update({ securityCode, requestId });
+    const dataOutput = data.request ? prettyRequestInfo(data) : `Verifying ${method.detail} as 2FA method`;
+    const text = 
+`
+NEAR Wallet security code: ${securityCode}\n\n
+Important: By entering this code, you are authorizing the following transaction:\n\n
+${dataOutput}
+`;
+    const html =
+`
+<body style="margin: 0; padding: 0;">
+    <table align="center" border="0" cellpadding="0" cellspacing="0" width="100%">
+        <tr>
+            <td align="center">
+                <img src="https://near.org/wp-content/themes/near-19/assets/downloads/near_logo.png" width="300" height="117"
+            </td>
+        </tr>
+        <tr>
+            <td>
+                <p>NEAR Wallet security code: ${securityCode}</p>
+                <p><strong>Important:</strong> By entering this code, you are authorizing the following transaction:\n\n</p>
+                <pre>
+                    ${dataOutput}
```

Comment:
let's make sure HTML cannot be injected here (e.g. by including HTML into function call data)

Reply:
Testing now and throwing 401 with message